cohortextractor/query_engines/mssql.py

```diff
@@ -73,22 +76,26 @@ def __init__(self, column_definitions, backend):
             )
 
         # Walk the nodes and identify output groups
-        self.output_groups = self.get_output_groups(column_definitions)
+        all_nodes = self.get_all_query_nodes(column_definitions)
+        self.output_groups = self.get_output_groups(all_nodes)
+        self.codelists = [node for node in all_nodes if isinstance(node, Codelist)]
```

Comment:
This seems to be operating at a different level of abstraction to `self.output_groups`, where the filtering gets its own method. Could we bung all this logic into the same method as `get_all_query_nodes()` and have it return multiple values?

Reply:
I don't quite follow this point here. I originally had this as a separate method for parity with `get_output_groups` but the code felt overall clearer to me having inlined the one-liner.

Reply:
I mean something like

```
self.output_groups, self.codelists = self.extract_nodes(column_definitions)
```

I don't feel very strongly.

Reply:
Ah, I see. I'm sure we can make the whole thing a bit clearer structurally. It might be better to see what else we end up wanting alongside codelists and output groups (there must be a better name for these!) before we do attempt that though.

Reply:
As you prefer. I'm secretly working up to trying to persuade you to completely restructure this "interpretation" of the DAG, so I hope we'll be revisiting this anyway.